src/modules/extra/m_utf8nick.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ * InspIRCd -- Internet Relay Chat Daemon
+ *
+ *   Copyright (C) 2015 Googolplexed <googol@googolplexed.net>
+ *
+ * This file is part of InspIRCd.  InspIRCd is free software: you can
+ * redistribute it and/or modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "inspircd.h"
+#include "caller.h"
+#include <unicode/unistr.h>
+#include <unicode/uniset.h>
+#include <unicode/uchar.h>
+#include <unicode/uscript.h>
+#include <unicode/ustring.h>
+
+#ifdef _WIN32
+#pragma comment(lib, "icuuc.lib")
+#endif
+
+//* $CompileFlags: pkgconfincludes("icu-uc","/unicode/ustring.h","") */
+/* $LinkerFlags: rpath("pkg-config --libs icu-uc") pkgconflibs("icu-uc","/libicuuc.so","-licuuc -licudata") */
+
+#define NICKBUFFERSIZE 512
+
+class ModuleUTF8 : public Module, public HandlerBase1<bool, const std::string&>
+{
+    caller1<bool, const std::string&> rememberer;
+    USet *validCharacters = NULL;
+public:
+    ModuleUTF8()
+            : rememberer(ServerInstance->IsNick) { }
+    //Check if Nick is valid using the current codepage
+    bool Call(const std::string& nick)
+    {
+        ServerInstance->Logs->Log(MODNAME, LOG_DEBUG, "Checking nick %s validity",nick.c_str());
+
+        //Nick length should be handled as per standard ASCII ? Not sure if this should be done pre or post conversion
+        if(nick.length() > ServerInstance->Config->Limits.NickMax)
+        {
+            return false;
+        }
+
+        char initialCharacter = nick[0];
+        if((initialCharacter <= '9' && initialCharacter >= '0') || initialCharacter == '-')
+        {
+            return false;
+        }
+
+        //Invalid input is replaced with U+FFFD
+        UChar unicodeString[NICKBUFFERSIZE];
+        int32_t outputLength;
+        UErrorCode err = U_ZERO_ERROR;
+        u_strFromUTF8(unicodeString,NICKBUFFERSIZE,&outputLength,nick.c_str(),-1,&err);
+        if(outputLength >= 128 || U_FAILURE(err))
```

Comment:
Using an uninitialized `outputLength` unless `u_strFromUTF8` guarantees that it writes a value to it in all cases